workspaces/local-cli/src/shared/coverage-2.ts

```diff
@@ -0,0 +1,129 @@
+
+      yield interactionToCoverage(spec, item.interaction.value)
+    }
+  })(captureIterator)
+
+  let combined_map: ICoverageMap = {};
+
+  for await(let individual_coverage of coverages){
+    for(const [k,v] of Object.entries(individual_coverage)) {
+      combined_map[k] = (combined_map[k]??0) + v;
+    }
+  }
+}
+
+type BodyDescriptor = {body?:{http_content_type: string, root_shape_id: string}};
+
+function interactionToCoverage(spec: DiffEngine.WasmSpecProjection, interaction: IHttpInteraction) {
+  const diff_str = DiffEngine.diff_interaction(JSON.stringify(interaction), spec);
+  // Maybe we could be more clever here later
+  // Would also love to get real types in here
+  const diff: Array<any> = JSON.parse(diff_str);
+
+  let is_diff = diff.length > 0;
+
+  let coverage_map: ICoverageMap = {};
+
+  recordEvent(TotalInteractions(), is_diff, coverage_map);
+
+  const path_id = DiffEngine.spec_resolve_path_id(spec, interaction.request.path);
+
+  if(path_id){
+    recordEvent(TotalForPath(path_id), is_diff, coverage_map);
+    const requests = JSON.parse(DiffEngine.spec_resolve_requests(spec, path_id, interaction.request.method)) as Array<[string,BodyDescriptor]>;
+    if(requests.length === 0) {
+      // Undocumented method
+    } else if(requests.length === 1) {
+      // Proceed
+      recordEvent(TotalForPathAndMethod(path_id,interaction.request.method), is_diff, coverage_map);
+      const [request_id, request] = requests[0];
+      recordEvent(TotalForRequest(request_id), is_diff, coverage_map);
+
+      if(request.body){
+        recordEvent(TotalForPathAndMethodAndContentType(path_id, interaction.request.method, request.body.http_content_type), is_diff, coverage_map);
+      } else {
+        recordEvent(TotalForPathAndMethodWithoutBody(path_id, interaction.request.method), is_diff, coverage_map);
+      }
+
+      const responses = JSON.parse(DiffEngine.spec_resolve_responses(spec, interaction.request.method, interaction.response.statusCode,path_id)) as Array<[string, BodyDescriptor]>;
+
+      if(responses.length === 0){
+        // Undocumented response
+      } else if(responses.length === 1) {
+        // Proceed
+        recordEvent(TotalForPathAndMethodAndStatusCode(path_id,interaction.request.method, interaction.response.statusCode), is_diff, coverage_map);
+        const [response_id, response] = responses[0];
+        recordEvent(TotalForResponse(response_id), is_diff, coverage_map);
+
+        if(response.body){
+          recordEvent(TotalForPathAndMethodAndStatusCodeAndContentType(path_id, interaction.request.method, interaction.response.statusCode, response.body.http_content_type), is_diff, coverage_map)
+        } else {
+          recordEvent(TotalForPathAndMethodAndStatusCodeWithoutBody(path_id, interaction.request.method, interaction.response.statusCode), is_diff, coverage_map)
+        }
+      } else {
+        // More than 1 response... Same as case below. Shouldn't happen I don't think?
```

Comment:
These things are technically possible when there are "domain conflicts" but safe to throw an error at each of these spots since it violates the invariants

Reply:
Okay great, so at least the places where I didn't know what logic to implement are the cases that we should never reach. Sweet!

Reply:
er...for a given (path_id, method), you can have 0+ bodies, aka 1+ requests each with a distinct content-type.

Similarly for a given (path_id, method, status_code) you can have 0+bodies aka 1+ responses each with a distinct content type. so I think you should check if length === 0 => error, otherwise, for each accumulate the counts.

Reply:
Thanks!